js/lib/components/feedback/selection.tsx

```diff
@@ -0,0 +1,98 @@
+import { VirtualElement } from "@popperjs/core";
+import React, { useCallback, useEffect, useState } from "react";
+import Highlighter from "web-highlighter";
+
+import { HIGHLIGHT_STORAGE_KEY } from "../../entryPoints/feedback";
+import FeedbackModal from "./modal";
+import SelectionTooltip from "./tooltip";
+
+type SelectionRendererProps = { highlighter: Highlighter; stored: any[] | undefined };
+let SelectionRenderer: React.FC<SelectionRendererProps> = ({ highlighter, stored }) => {
+  // id of feedback highlight currently being hovered over
+  const [hovered, setHovered] = useState<false | string>(false);
+
+  // current highlighted range of text
+  const [currRange, setCurrRange] = useState<false | Range>(false);
+
+  // whether feedback modal is open
+  const [modalOpen, setModalOpen] = useState(false);
+
+  // update selected range when selection changes
+  const handleSelection = useCallback(() => {
+    // get current selection (falsy value if no selection)
+    let selection = document.getSelection();
+    let range =
+      selection && !selection.isCollapsed && selection.rangeCount && selection.getRangeAt(0);
+
+    if (range) {
+      setCurrRange(range);
+    } else {
+      setCurrRange(false);
+    }
+  }, []);
+
+  useEffect(() => {
+    // load highlights from local storage
+    stored?.map(s => highlighter.fromStore(s.startMeta, s.endMeta, s.id, s.text, s.extra));
+
+    // store new highlights in localStorage when created
+    highlighter.on(Highlighter.event.CREATE, ({ sources }) => {
+      let stored_str = localStorage.getItem(HIGHLIGHT_STORAGE_KEY);
+      let stored_highlights = JSON.parse(stored_str || "[]");
+      stored_highlights.push(...sources);
+
+      localStorage.setItem(HIGHLIGHT_STORAGE_KEY, JSON.stringify(stored_highlights));
+    });
+
+    // update state on hover changes
+    highlighter.on(Highlighter.event.HOVER, ({ id }) => setHovered(id));
+    highlighter.on(Highlighter.event.HOVER_OUT, () => setHovered(false));
+  }, []);
+
+  useEffect(() => {
+    // handle selection events only when modal closed
+    if (!modalOpen) {
+      document.addEventListener("selectionchange", handleSelection);
+    } else {
+      document.removeEventListener("selectionchange", handleSelection);
+    }
+  }, [modalOpen]);
+
+  // Remove modal and tooltip when closing modal
+  const handleCloseModal = () => {
+    setModalOpen(false);
+    setCurrRange(false);
+  };
+
+  if (hovered) {
```

Comment:
I feel like this renderer is a fairly complex state machine which could be factored a bit. Like this logic for showing the feedback on hover could be its own component, it doesn't need to be tied to the logic that shows a tooltip for selected text.